parlai/tasks/insuranceqa/build.py

```diff
@@ -11,99 +11,173 @@
-            bad_ans = [d_label_answer[aid_] for aid_ in s_bad_aids.split()]
-            # save
-            s = '1 ' + q + '\t' + "|".join(good_ans) + '\t\t' + "|".join(good_ans + bad_ans)
-            fout.write(s + '\n')
-    fout.close()
+        # the root of dataset
+        dpext = os.path.join(dpath, 'insuranceQA-master/%s' % cls.version)
+
+        # read vocab file
+        vocab_path = os.path.join(dpext, "vocabulary")
+        d_vocab = cls.read_vocab(vocab_path)
+
+        # read label2answer file
+        label2answer_path_gz = os.path.join(dpext, cls.label2answer_fname)
+        d_label_answer = cls.read_label2answer(label2answer_path_gz, d_vocab)
+
+        # Create out path
+        out_path = os.path.join(dpath, cls.version)
+        build_data.make_dir(out_path)
+
+        # Parse and write data files
+        for dtype, data_fname in cls.data_fnames:
+            data_path = os.path.join(dpext, data_fname)
+            cls.create_fb_format(out_path, dtype, data_path, d_vocab, d_label_answer)
+
+
+class ParseInsuranceQAV1(ParseInsuranceQA):
+    version = "V1"
+    label2answer_fname = "answers.label.token_idx"
+    data_fnames = [
+        ("train", "question.train.token_idx.label"),
+        ("valid", "question.dev.label.token_idx.pool"),
+        ("test", "question.test1.label.token_idx.pool"),
+        # ("test2", "question.test2.label.token_idx.pool")
+    ]
+
+    @classmethod
+    def create_fb_format(cls, out_path, dtype, inpath, d_vocab, d_label_answer):
+        print('building fbformat:' + dtype)
+        fout = open(os.path.join(out_path, dtype + '.txt'), 'w')
+        lines = open(inpath).readlines()
+
+        for line in lines:
+            fields = line.rstrip("\n").split("\t")
+            if dtype == "train":
+                assert len(fields) == 2, "data file (%s) corrupted." % inpath
+                s_q_wids, s_good_aids = fields
+
+                q = cls.wids2sent(s_q_wids.split(), d_vocab)
+                good_ans = [d_label_answer[aid_] for aid_ in s_good_aids.split()]
+                # save good answers (train only)
+                s = '1 ' + q + '\t' + "|".join(good_ans)
+                fout.write(s + '\n')
+            else:
+                assert len(fields) == 3, "data file (%s) corrupted." % inpath
+                s_good_aids, s_q_wids, s_bad_aids = fields
+
+                q = cls.wids2sent(s_q_wids.split(), d_vocab)
+                good_ans = [d_label_answer[aid_] for aid_ in s_good_aids.split()]
+                bad_ans = [d_label_answer[aid_] for aid_ in s_bad_aids.split()]
+                # save good answers and candidates
+                s = '1 ' + q + '\t' + "|".join(good_ans) + '\t\t' + "|".join(good_ans + bad_ans)
+                fout.write(s + '\n')
+        fout.close()
+
+
+class ParseInsuranceQAV2(ParseInsuranceQA):
+    version = "V2"
+    label2answer_fname = "InsuranceQA.label2answer.token.encoded.gz"
+    # TODO: support 500, 1000, 1500 (# of label candidates)
```

Comment:
could have the build file set up all of them, and then the teacher in the agent file decide which one to use

for example, the babi tasks have a "Task1kTeacher" and "Task10kTeacher" for the two different version, but you can specify specific tasks with "babi:Task1k:1" or "babi:task10k:2".